clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -394,14 +392,14 @@ public void failUnsentRequests(Node node, RuntimeException e) {
     private boolean trySend(long now) {
         // send any requests that can be sent now
         boolean requestsSent = false;
-        for (Map.Entry<Node, List<ClientRequest>> requestEntry: unsent.entrySet()) {
-            Node node = requestEntry.getKey();
-            Iterator<ClientRequest> iterator = requestEntry.getValue().iterator();
-            while (iterator.hasNext()) {
-                ClientRequest request = iterator.next();
-                if (client.ready(node, now)) {
+        Set<Node> nodes = unsent.getNodes();
+        for (Node node: nodes) {
+            if (client.ready(node, now)) {
+                // Remove entry before invoking request callback to avoid callbacks handling
+                // coordinator failures traversing the unsent list again.
+                List<ClientRequest> requests = unsent.remove(node);
```

Comment:
It is possible to provide a method `clear(node)` that returns a list of request corresponding to the node and keep the node as key. But I am not sure we can get an performance benefits by making this change. As of now every `trySend()` may cause a key to be deleted and added later, and we need to create an empty ArrayList when adding that key. If we make that change using `clear()`, every `trySend()` may require us to create a new ArrayList() that contains every element of original list of requests.